services/sinner/contracts/origin/parser.go

```diff
@@ -0,0 +1,164 @@
+// Package origin is the origin contract parser.
+package origin
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/synapsecns/sanguine/services/sinner/logger"
+
+	"github.com/synapsecns/sanguine/agents/contracts/origin"
+	"github.com/synapsecns/sanguine/agents/types"
+	"github.com/synapsecns/sanguine/services/sinner/db"
+	"github.com/synapsecns/sanguine/services/sinner/db/model"
+	sinnerTypes "github.com/synapsecns/sanguine/services/sinner/types"
+)
+
+// ParserImpl is the parser for the origin contract.
+type ParserImpl struct {
+	filterer *origin.OriginFilterer
+	// parser is the parser interface.
+	parser origin.Parser
+	// db is the database
+	db db.EventDB
+	// txMap is a map of tx hashes to tx data. Exported for testing.
+	txMap map[string]sinnerTypes.TxSupplementalInfo
+	// chainID is the chainID of the underlying chain
+	chainID uint32
+}
+
+// NewParser creates a new parser for the origin contract.
+func NewParser(originAddress common.Address, db db.EventDB, chainID uint32) (*ParserImpl, error) {
+	// Get agents parser to utilize event type parsing.
+	agentsParser, err := origin.NewParser(originAddress)
+	if err != nil {
+		return nil, fmt.Errorf("could not create agents parser %w", err)
+	}
+
+	// Get filterer to get ABI IFace from abi.
+	filter, err := origin.NewOriginFilterer(originAddress, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not create %T: %w", origin.OriginFilterer{}, err)
+	}
+
+	parser := &ParserImpl{
+		filterer: filter,
+		parser:   agentsParser,
+		db:       db,
+		chainID:  chainID,
+	}
+	return parser, nil
+}
+
+// UpdateTxMap updates the tx map so that scribe does not have to be requested for each log.
+// This function is not concurrency safe, and is intended to be used before using ParseAndStore.
+func (p *ParserImpl) UpdateTxMap(txMap map[string]sinnerTypes.TxSupplementalInfo) {
+	p.txMap = txMap
+}
+
+// UnsafeGetTXMap gets the tx map strictly for testing purposes.
+func (p *ParserImpl) UnsafeGetTXMap() map[string]sinnerTypes.TxSupplementalInfo {
+	return p.txMap
+}
+
+// ParseAndStore parses and stores the log.
+func (p *ParserImpl) ParseAndStore(ctx context.Context, log ethTypes.Log) error {
```

Comment:
I think this makes sense, the decision to do this was due to there being only one to two events max per contract, leading to lots of new small files and folders w/similar structure to the contract module if a parser module was created.

Saw a similar pattern in agents specifically for parsing so used that as a part of my decision making process, but I do think it might be easier for maintainability down the road to have a parser module. Happy to change that in this PR 👍